File: packages/dart/lib/Seguranca/Ataques/AtaqueDistribuicaoBytes.dart
import 'dart:math';

import 'package:criptografia/Seguranca/AlvoCriptografico.dart';
import 'package:criptografia/Seguranca/Ataque.dart';
import 'package:criptografia/Seguranca/ResultadoAtaque.dart';
import 'package:criptografia/Seguranca/Util.dart';

/// Um keystream de qualidade deve ter bytes distribuídos uniformemente entre
/// 0-255. Um viés forte (qui-quadrado muito alto) indica que alguns valores
/// de byte saem com mais frequência que outros - sinal de fraqueza estatística
/// na função de mistura.
class AtaqueDistribuicaoBytes extends Ataque {
  final int _amostrasDeBlocos;
  final int _tamanhoBloco;

  AtaqueDistribuicaoBytes({
    this._amostrasDeBlocos = 500,
    this._tamanhoBloco = 32,
  });

  @override
  String nome() => 'Distribuição de bytes (qui-quadrado)';

  @override
  ResultadoAtaque executar(AlvoCriptografico alvo) {
    final chave = alvo.chaveDeTeste();

    final contagem = List<int>.filled(256, 0);
    var total = 0;
    for (var i = 0; i < _amostrasDeBlocos; i++) {
      final ks = alvo.gerarKeystreamBruto(
        chave,
        randomBytes(alvo.tamanhoIv()),
        'enc',
        _tamanhoBloco,
      );
      for (var j = 0; j < ks.length; j++) {
        contagem[ks[j]]++;
        total++;
      }
    }

    final esperado = total / 256;
    var qui2 = 0.0;
    for (final c in contagem) {
      qui2 += pow(c - esperado, 2).toDouble() / esperado;
    }

    // Com 255 graus de liberdade, valores acima de ~330 já são
    // estatisticamente suspeitos (p < 0.01); acima de ~380, bem suspeitos.
    final vulneravel = qui2 > 330;

    return ResultadoAtaque(
      nome(),
      vulneravel,
      vulneravel ? Severidade.media : Severidade.info,
      'qui-quadrado=${qui2.toStringAsFixed(1)} sobre $total bytes '
          '(255 graus de liberdade; >330 é suspeito)',
      {'qui2': qui2},
    );
  }
}

File: packages/julia/src/Seguranca/Ataques/AtaqueDistribuicaoBytes.jl
"""
Um keystream de qualidade deve ter bytes distribuídos uniformemente entre
0-255. Um viés forte (qui-quadrado muito alto) indica que alguns valores
de byte saem com mais frequência que outros - sinal de fraqueza estatística
na função de mistura.
"""
struct AtaqueDistribuicaoBytes <: Ataque
    amostras_de_blocos::Int
    tamanho_bloco::Int
end

AtaqueDistribuicaoBytes(;
    amostras_de_blocos::Integer = 500,
    tamanho_bloco::Integer = 32,
) = AtaqueDistribuicaoBytes(Int(amostras_de_blocos), Int(tamanho_bloco))

nome(::AtaqueDistribuicaoBytes) = "Distribuição de bytes (qui-quadrado)"

function executar(a::AtaqueDistribuicaoBytes, alvo::AlvoCriptografico)
    chave = chave_de_teste(alvo)

    contagem = zeros(Int, 256)
    total = 0
    for _ in 1:(a.amostras_de_blocos)
        ks = gerar_keystream_bruto(
            alvo,
            chave,
            random_bytes(tamanho_iv(alvo)),
            "enc",
            a.tamanho_bloco,
        )
        for byte in ks
            contagem[byte+1] += 1
            total += 1
        end
    end

    esperado = total / 256
    qui2 = 0.0
    for c in contagem
        qui2 += (c - esperado)^2 / esperado
    end

    # Com 255 graus de liberdade, valores acima de ~330 já são
    # estatisticamente suspeitos (p < 0.01); acima de ~380, bem suspeitos.
    vulneravel = qui2 > 330

    return ResultadoAtaque(
        nome(a),
        vulneravel,
        vulneravel ? "media" : "info",
        "qui-quadrado=$(fmt(qui2, 1)) sobre $total bytes (255 graus de liberdade; >330 é suspeito)",
        Dict{String,Any}("qui2" => qui2),
    )
end
